### PR TITLE
feat: add matrax environment

### PR DESCRIPTION
**Pull request by JemmaLDaniel**
## What?
Added support for Matrax via a wrapper and a config file.
## Why?
This makes Mava compatible with Matrax.
## How?
The Matrax wrapper transforms observations to the standard Mava API, and the config allows us to choose between and run Matrax's Climbing, Penality, Conflict and NoConflict games with Mava's implemented MARL algorithms.

### CODE REVIEW

File: mava/configs/env/matrax.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+  - _self_
+
+env_name: Matrax
+
+scenario:
+  task_name: Climbing-stateless-v0
+
+eval_metric: episode_return
+
+# Possible scenarios:
```
Comment (RuanJohn) [inline mava/configs/env/matrax.yaml:7]: NIT: Please move this up to be above `scenario` to be consistent with the other env configs.

File: mava/utils/make_env.py [hunk 2]
```diff
@@ -124,6 +131,38 @@ def make_jaxmarl_env(
     return env, eval_env
 
 
+def make_matrax_env(
```
Comment (RuanJohn) [inline mava/utils/make_env.py:139]: Very nice! :sunglasses:

File: mava/wrappers/matrax.py [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        )
+
+    def modify_timestep_reset(self, timestep: TimeStep) -> TimeStep[Observation]:
+        """Modify the timestep for `step` and `reset`."""
+        observation = Observation(
+            agents_view=timestep.observation.agent_obs,
+            action_mask=jnp.ones((self._num_agents, self._num_actions), dtype=bool),
+            step_count=jnp.repeat(timestep.observation.step_count, self._num_agents),
+        )
+        reward = jnp.repeat(timestep.reward, self._num_agents)
```
Comment (RuanJohn) [inline mava/wrappers/matrax.py]: Can you please install the latest version of Matrax and double check that this still runs? I am pretty sure we fixed this so that the reset won't require reward duplication anymore.

File: mava/wrappers/matrax.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    def modify_timestep_step(self, timestep: TimeStep) -> TimeStep[Observation]:
+        """Modify the timestep for `step` and `reset`."""
+        observation = Observation(
+            agents_view=timestep.observation.agent_obs,
+            action_mask=jnp.ones((self._num_agents, self._num_actions), dtype=bool),
+            step_count=jnp.repeat(timestep.observation.step_count, self._num_agents),
+        )
+        return timestep.replace(
+            observation=observation,
+            reward=timestep.reward * 1.0,
```
Comment (RuanJohn) [inline mava/wrappers/matrax.py]: Is the `* 1.0` to convert the reward to a float?

File: mava/wrappers/matrax.py [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            jnp.bool_,
+            "action_mask",
+        )
+        return specs.Spec(
+            Observation,
+            "ObservationSpec",
+            agents_view=env_observation_spec,
+            action_mask=action_mask,
+            step_count=step_count,
+        )
```
Comment (RuanJohn) [inline mava/wrappers/matrax.py]: Since we only use this once, we can just define it in the returned `Spec`.

```suggestion
        """Specification of the observation of the environment."""
        step_count = specs.BoundedArray(
            (self._num_agents,),
            jnp.int32,
            [0] * self._num_agents,
            [self.time_limit] * self._num_agents,
            "step_count",
        )
        action_mask = specs.Array(
            (self._num_agents, self._num_actions),
            jnp.bool_,
            "action_mask",
        )
        return specs.Spec(
            Observation,
            "ObservationSpec",
            agents_view=self._env.observation_spec().agent_obs,
            action_mask=action_mask,
            step_count=step_count,
        )
```

File: requirements/requirements.txt [hunk 6]
```diff
@@ -7,6 +7,7 @@ jax
 jaxlib
 jaxmarl
 jumanji @ git+https://github.com/sash-a/jumanji
+matrax
```
Comment (RuanJohn) [inline requirements/requirements.txt]:
```suggestion
matrax @ git+https://github.com/instadeepai/matrax
```

File: mava/wrappers/matrax.py (hunk 3 above)
Comment (RuanJohn) [reply mava/wrappers/matrax.py]: This should also allow you to only have single `modify_timstep` method that should work for both `step` and `reset`.

File: mava/utils/make_env.py [hunk 7]
```diff
@@ -124,6 +131,38 @@ def make_jaxmarl_env(
...
+    """
+    Creates Matrax environments for training and evaluation.
+
+    Args:
+        env_name (str): The name of the environment to create.
+        config (Dict): The configuration of the environment.
+        add_global_state (bool): Whether to add the global state to the observation.
+
+    Returns:
+        A tuple of the Matrax environments.
```
Comment (RuanJohn) [inline mava/utils/make_env.py]: NIT: 
```suggestion
        A tuple containing a train and evaluation Matrax environment.
```

File: mava/utils/make_env.py [hunk 8]
```diff
@@ -124,6 +131,38 @@ def make_jaxmarl_env(
...
 
 
+def make_matrax_env(
+    env_name: str, config: DictConfig, add_global_state: bool = False
+) -> Tuple[Environment, Environment]:
+    """
+    Creates Matrax environments for training and evaluation.
+
+    Args:
+        env_name (str): The name of the environment to create.
```
Comment (RuanJohn) [inline mava/utils/make_env.py]: NIT: 
```suggestion
        env_name: The name of the environment to create.
```

File: mava/utils/make_env.py [hunk 9]
```diff
@@ -124,6 +131,38 @@ def make_jaxmarl_env(
...
+        config (Dict): The configuration of the environment.
+        add_global_state (bool): Whether to add the global state to the observation.
```
Comment (RuanJohn) [inline mava/utils/make_env.py]: NIT: 
```suggestion
        config: The configuration of the environment.
        add_global_state: Whether to add the global state to the observation.
```

**Review by RuanJohn — CHANGES_REQUESTED** [review]
Thank you for this @JemmaLDaniel! A very nice first PR into Mava :fire: 
Just a few minor comments from my side.

File: mava/utils/make_env.py [hunk 10]
```diff
@@ -42,6 +44,11 @@
     "LevelBasedForaging-v0": {"generator": LbfRandomGenerator, "wrapper": LbfWrapper},
 }
 
+# Define a different registry for Matrax since it has no generator.
+_matrax_registry = {
+    "Matrax": {"wrapper": MatraxWrapper},
+}
```
Comment (sash-a) [inline mava/utils/make_env.py]: Because we only have 1 wrapper we don't need a registry for matrax :smile: 
```suggestion
```

File: mava/utils/make_env.py [hunk 11]
```diff
@@ -124,6 +131,38 @@ def make_jaxmarl_env(
...
+        A tuple of the Matrax environments.
+    """
+    # Select the Matrax wrapper.
+    wrapper = _matrax_registry[env_name]["wrapper"]
+
+    # Create envs.
+    task_name = config["env"]["scenario"]["task_name"]
+    env = matrax.make(task_name, **config.env.kwargs)
+    eval_env = matrax.make(task_name, **config.env.kwargs)
+    env, eval_env = wrapper(env), wrapper(eval_env)
```
Comment (sash-a) [inline mava/utils/make_env.py]: We can just use the wrapper directly here

```suggestion
    # Create envs.
    task_name = config["env"]["scenario"]["task_name"]
    env = matrax.make(task_name, **config.env.kwargs)
    eval_env = matrax.make(task_name, **config.env.kwargs)
    env, eval_env = MatraxWrapper(env), MatraxWrapper(eval_env)
```

File: mava/utils/make_env.py [hunk 12]
```diff
@@ -141,5 +180,7 @@ def make(config: DictConfig, add_global_state: bool = False) -> Tuple[Environmen
         return make_jumanji_env(env_name, config, add_global_state)
     elif env_name in jaxmarl.registered_envs:
         return make_jaxmarl_env(env_name, config, add_global_state)
+    elif env_name in _matrax_registry:
```
Comment (sash-a) [inline mava/utils/make_env.py:188]:
```suggestion
    elif env_name == "Matrax":
```

File: mava/utils/make_env.py (hunk 10 above)
Comment (sash-a) [reply mava/utils/make_env.py]: I think the `make_matrax_env` and the `make_jumanji_env` methods are basically the same, so what we could do is join the matrax and jumanji registries and then just use the `make_jumanji_env` methods. 

@RuanJohn @JemmaLDaniel what do you think?

File: mava/wrappers/matrax.py (hunk 4 above)
Comment (sash-a) [reply mava/wrappers/matrax.py]: if so: `timestep.reward.astype(float)` is more clear

File: mava/wrappers/matrax.py [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        step_count = specs.BoundedArray(
+            (self._num_agents,),
+            jnp.int32,
+            [0] * self._num_agents,
+            [self.time_limit] * self._num_agents,
+            "step_count",
+        )
+        action_mask = specs.Array(
+            (self._num_agents, self._num_actions),
+            jnp.bool_,
```
Comment (sash-a) [inline mava/wrappers/matrax.py]:
```suggestion
            bool,
```

**Review by sash-a — DISMISSED** [review]
Awesome first PR! Looking almost ready to go :fire: :rocket:

File: mava/utils/make_env.py (hunk 10 above)
Comment (JemmaLDaniel) [reply mava/utils/make_env.py]: That is definitely more concise, but I think keeping them separate is slightly easier to read and understand because it is in keeping with the choice for Matrax as a separate suite of environments to Jumanji. Another solution is to merge them into one registry but change the name to be more all-encompassing than `make_jumanji_env` and `_jumanji_registry`?

File: mava/wrappers/matrax.py (hunk 3 above)
Comment (JemmaLDaniel) [reply mava/wrappers/matrax.py]: Fixed--thanks!

File: mava/wrappers/matrax.py (hunk 4 above)
Comment (JemmaLDaniel) [reply mava/wrappers/matrax.py]: It was, but it's no longer necessary with the updated Matrax install 😄

File: mava/utils/make_env.py (hunk 12 above)
Comment (JemmaLDaniel) [reply mava/utils/make_env.py:188]: It seems like this change and the previous one to use `MatraxWrapper` directly make `_matrax_registry` completely unnecessary. Could we leave it out altogether?

File: mava/utils/make_env.py (hunk 10 above)
Comment (RuanJohn) [reply mava/utils/make_env.py]: I agree that the methods are essentially the same, but I think it is better to keep them separate and to have a matrax registry. The reason for this is that the poly games may become part of matrax later on keeping env suites separate from each other sets a nice pattern.

Comment (sash-a) [reply mava/utils/make_env.py]: Ok sure happy to keep with this pattern :smile:

File: mava/utils/make_env.py (hunk 12 above)
Comment (sash-a) [reply mava/utils/make_env.py:188]: Ye I think leave it in for now, just like @RuanJohn said it keeps the pattern

File: mava/wrappers/matrax.py [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+        self._num_agents = self._env.num_agents
+        self.time_limit = self._env.time_limit
+        self._num_actions = self._env.num_actions
+        self.num_obs_features = self._env.observation_spec().agent_obs.shape[-1]
+
+    def modify_timestep(self, timestep: TimeStep) -> TimeStep[Observation]:
+        """Modify the timestep for `step` and `reset`."""
+        observation = Observation(
+            agents_view=timestep.observation.agent_obs,
+            action_mask=jnp.ones((self._num_agents, self._num_actions), dtype=bool),
```
Comment (WiemKhlifi) [inline mava/wrappers/matrax.py]: Since this is fixed, can we move it up so we don't have to calculate it over and over at each timestep:
```suggestion
            action_mask=self.action_mask

```

File: mava/wrappers/matrax.py [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+from mava.types import Observation, State
+
+
+class MatraxWrapper(Wrapper):
+    def __init__(self, env: Environment):
+        super().__init__(env)
+        self._num_agents = self._env.num_agents
+        self.time_limit = self._env.time_limit
+        self._num_actions = self._env.num_actions
+        self.num_obs_features = self._env.observation_spec().agent_obs.shape[-1]
```
Comment (WiemKhlifi) [inline mava/wrappers/matrax.py]: Same,  the `num_obs_features` isn't used here can you remove it please and add this as discussed in the suggestion above:
```suggestion
        self.action_mask = jnp.ones((self._num_agents, self._num_actions), dtype=bool)
```

File: mava/wrappers/matrax.py [hunk 16]
```diff
@@ -0,0 +1,78 @@
...
+        state, timestep = self._env.step(state, action)
+        return state, self.modify_timestep(timestep)
+
+    def observation_spec(self) -> specs.Spec[Observation]:
+        """Specification of the observation of the environment."""
+        step_count = specs.BoundedArray(
+            (self._num_agents,),
+            jnp.int32,
+            [0] * self._num_agents,
+            [self.time_limit] * self._num_agents,
```
Comment (WiemKhlifi) [inline mava/wrappers/matrax.py]:
``` suggestion
            [self._env.time_limit] * self._num_agents,
```

File: mava/wrappers/matrax.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+from jumanji.wrappers import Wrapper
+
+from mava.types import Observation, State
+
+
+class MatraxWrapper(Wrapper):
+    def __init__(self, env: Environment):
+        super().__init__(env)
+        self._num_agents = self._env.num_agents
+        self.time_limit = self._env.time_limit
```
Comment (WiemKhlifi) [inline mava/wrappers/matrax.py]: Can we access it directly in the spec, since we won't have to use it again
``` suggestion

```

File: mava/wrappers/matrax.py [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+import jax.numpy as jnp
+from jumanji import specs
+from jumanji.env import Environment
+from jumanji.types import TimeStep
+from jumanji.wrappers import Wrapper
+
+from mava.types import Observation, State
+
+
+class MatraxWrapper(Wrapper):
```
Comment (WiemKhlifi) [inline mava/wrappers/matrax.py:27]:
``` suggestion
class MatraxWrapper(Wrapper):
    """Multi-agent wrapper for the Matrax environment."""

```

File: mava/utils/make_env.py [hunk 19]
```diff
@@ -131,6 +138,38 @@ def make_jaxmarl_env(
...
+    Args:
+        env_name: The name of the environment to create.
+        config: The configuration of the environment.
+        add_global_state: Whether to add the global state to the observation.
+
+    Returns:
+        A tuple containing a train and evaluation Matrax environment.
+    """
+    # Select the Matrax wrapper.
+    wrapper = _matrax_registry[env_name]["wrapper"]
```
Comment (WiemKhlifi) [inline mava/utils/make_env.py]:
``` suggestion
    wrapper = _matrax_registry[env_name]
```

File: mava/utils/make_env.py [hunk 20]
```diff
@@ -43,6 +45,11 @@
     "LevelBasedForaging-v0": {"generator": LbfRandomGenerator, "wrapper": LbfWrapper},
 }
 
+# Define a different registry for Matrax since it has no generator.
+_matrax_registry = {
+    "Matrax": {"wrapper": MatraxWrapper},
+}
```
Comment (WiemKhlifi) [inline mava/utils/make_env.py]: Small suggestion here since we only using a wrapper (no generators like in jumanji):
``` suggestion
_matrax_registry = {
    "Matrax": {MatraxWrapper},
}
```

**Review by WiemKhlifi — CHANGES_REQUESTED** [review]

File: mava/configs/env/matrax.yaml [hunk 21]
```diff
@@ -0,0 +1,26 @@
...
+#   3. NoConflict-{id}-{state}-v0
+#       3.1. id = {1, 2, 3, ..., 20, 21}
+#       3.2. state = {stateful, stateless}
+#   4. Conflict-{id}-{state}-v0
+#       4.1. id = {1, 2, 3, ..., 56, 57}
+#       4.2. state = {stateful, stateless}
+scenario:
+  task_name: Climbing-stateless-v0
+  name: Matrax
+
```
Comment (WiemKhlifi) [inline mava/configs/env/matrax.yaml:22]:
``` suggestion
# Defines the metric that will be used to evaluate the performance of the agent.

```

File: mava/utils/make_env.py (hunk 20 above)
Comment (JemmaLDaniel) [reply mava/utils/make_env.py]: I agree, makes more sense with the separately defined registries to tailor it more to Matrax 😄

**Review by WiemKhlifi — APPROVED** [review]
Very clean PR! Thank you @JemmaLDaniel 🔥🚀

**Review by RuanJohn — APPROVED** [review]
Epic @JemmaLDaniel :fire:
